manifests/agent/windows.pp

```diff
@@ -12,62 +17,74 @@
     $url = "https://s3.amazonaws.com/download.distelli.com/distelli.Windows-x86/${archive}"
   }
 
-  $workdir = 'C:/Program Files/Distelli'
+  if $pipelines::agent::user_groups {
+    $user_groups = ['Users','Administrators'] + $pipelines::agent::user_groups
+  }
+  else {
+    $user_groups = ['Users','Administrators']
+  }
+
+  user { 'distelli':
+    ensure     => present,
+    comment    => 'Puppet Pipelines User',
+    home       => $pipelines::agent::user_home,
+    groups     => $user_groups,
+    password   => $pipelines::agent::user_password,
+    managehome => true,
+  }
 
-  archive { "${tempdir}/${archive}" :
+  archive { "${tempdir}/${archive}":
     source       => $url,
-    # cleanup      => false,
     creates      => "${tempdir}/distelli",
     extract      => true,
     extract_path => $tempdir,
     require      => User['distelli'],
   }
 
-  file { $workdir :
-    ensure  => directory,
-    owner   => 'distelli',
-    group   => 'Administrators',
-    require => User['distelli'],
+  file { $workdir:
+    ensure => directory,
+    owner  => 'distelli',
+    group  => 'Administrators',
   }
 
-  # Requires fqdn_rand_string function from puppetlabs/stdlib
-  $distelli_exec = "distelli-${fqdn_rand_string(5)}.exe"
-
-  exec { 'Copy executable' :
+  exec { 'pipelines::agent::windows Copy executable':
     command   => "Copy-Item ${tempdir}/distelli \$ENV:ProgramFiles/Distelli/${distelli_exec}",
-    unless    => "If (Test-Path -Path \$ENV:ProgramFiles/Distelli/${distelli_exec}) { exit 0 } else { exit 1}",
+    unless    => "If (Test-Path -Path \$ENV:ProgramFiles/Distelli/${distelli_exec}) { exit 0 } else { exit 1 }",
     provider  => powershell,
-    require   => File[$workdir],
     logoutput => true,
+    require   => File[$workdir],
+    subscribe => Archive["${tempdir}/${archive}"],
   }
 
-  file { ["${workdir}/distelli.exe",  "${workdir}/dagent.exe", "${workdir}/dtk.exe"] :
+  file { ["${workdir}/distelli.exe",  "${workdir}/dagent.exe", "${workdir}/dtk.exe"]:
     ensure  => link,
     target  => "${workdir}/${distelli_exec}",
-    require => Exec['Copy executable'],
+    require => Exec['pipelines::agent::windows Copy executable'],
   }
 
-  exec { 'Test distelli.exe execution' :
-    command     => "& \$ENV:ProgramFiles/Distelli/distelli.exe version",
+  exec { 'pipelines::agent::windows Test distelli.exe execution':
+    command     => '& $ENV:ProgramFiles/Distelli/distelli.exe version',
     provider    => powershell,
-    subscribe   => File["${workdir}/distelli.exe"],
     logoutput   => true,
     refreshonly => true,
+    subscribe   => File["${workdir}/distelli.exe"],
   }
 
-  file { 'C:/distelli.yml' :
-    ensure  => file,
-    owner   => 'distelli',
-    group   => 'Administrators',
-    mode    => '0644',
-    content => epp('distelli/distelli.yml.epp'),
-    require => Exec['Test distelli.exe execution'],
+  file { 'C:/distelli.yml':
```

Comment:
Dangerous Default. Do not hardcode system drive

Also, this is not the location for application data.  That should be  `%APPDATA%\PuppetLabs\pipelines` or similar.

Reply:
Application data: This has a similar problem on Linux; it installs the binary in `/home/distelli`. :( I reverted my change to make it `/opt/distelli` because every time something broke that was always the first thing people tried to revert, and it's a pain to change. 😭

Reply:
Does the Distelli agent itself expect this file to be here? If so, hands may be tied?

Reply:
I'm not sure about Windows. On Linux and macOS, it works with a different location, but it's not the default when you use the installer.

Reply:
I would love to make this follow common practice if not standards. I'm just running into practical problems, and trying to avoid them to get this merged.

If y'all think it's better to change the defaults to something more standard, I would be very happy to do that.

Reply:
So in Windows there's a separation between application executables and application data (Program Files vs App Data)

From the distelli docs
https://pipelines.puppet.com/docs/agent/automated-install-of-the-distelli-agent/
```
Placing the File
The default location for placing the distelli.yml file is /etc.

On Windows, the default location is %SystemDrive%\distelli.yml.

You can override the default location with the -conf option.
```

IMO using %SYSTEMDRIVE% was a quick and dirty method, really should be using AppData (That's been around since WinXP/Server 2003 days)

Reply:
Unfortunately I don't see a good way to change the configuration path from `%SystemDrive%\distelli.yml`, since it's baked into the distelli executable (I assume). I am changing it to use `%SystemDrive%`, though.